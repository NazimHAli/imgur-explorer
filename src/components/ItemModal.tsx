import { ItemModalComments } from "@/components/ItemModalComments";
import { useGlobalContext } from "@/state/GlobalContext";
import { truncateText } from "@/utils/dataUtils";
import { Dispatch, SetStateAction, useEffect } from "react";
import { KeyboardEvent, MouseEvent } from "react";
import { ThumbsUp, MessageSquare, Eye, Icon } from "react-feather";
import Modal from "react-modal";

// Bind modal to appElement for accessibility
Modal.setAppElement("#root");

function iconWithDataBadge(dataCount: number, TheIcon: Icon): JSX.Element {
  return (
    <span className="data-badge" data-count={dataCount?.toLocaleString()}>
      <TheIcon width="20" height="20" />
    </span>
  );
}

function ItemModal(props: {
  isOpen: boolean;
  setIsOpen: Dispatch<SetStateAction<boolean>>;
}): JSX.Element {
  const { isOpen, setIsOpen } = props;
  const { setState, state } = useGlobalContext();

  function closeModal(
    event: MouseEvent<Element, globalThis.MouseEvent> | KeyboardEvent<Element>
  ): void {
    setIsOpen(false);
    setState((currentState) => {
      return { ...currentState, selectedItem: {}, selectedItemComments: [] };
    });
    event.preventDefault();
  }

  function openModal(): void {
    setIsOpen(true);
  }

  useEffect(() => {
    if (state.selectedItemComments.length > 0) {
      openModal();
    }
  }, [state.selectedItem, state.selectedItemComments]);

  return (
    <Modal
      contentLabel="Image Modal"
      className="myModal__Content"
      overlayClassName="myModal__Overlay"
      isOpen={isOpen}
      onRequestClose={closeModal}
      preventScroll={true}
    >
      {state.selectedItem?.images && (
        <div className="item-modal">
          <h3 className="item-modal__title">
            Title: {truncateText(state.selectedItem?.title, 100)}
          </h3>

          <button className="item-modal--button__close" onClick={closeModal}>
            <svg
              xmlns="https://www.w3.org/2000/svg"
              viewBox="0 0 20 20"
              fill="currentColor"
            >
              <path
                fillRule="evenodd"
                d="M4.293 4.293a1 1 0 011.414 0L10 8.586l4.293-4.293a1 1 0 111.414 1.414L11.414 10l4.293 4.293a1 1 0 01-1.414 1.414L10 11.414l-4.293 4.293a1 1 0 01-1.414-1.414L8.586 10 4.293 5.707a1 1 0 010-1.414z"
                clipRule="evenodd"
              />
            </svg>
          </button>

          <div className="item-modal__image">
            <img
              alt={state.selectedItem?.title}
              width={state.selectedItem?.images[0].width}
              height={state.selectedItem?.images[0].height}
              srcSet={state.selectedItem?.images[0].link}
              loading="lazy"
            />
          </div>

          <div className="card-info__icons">
            {iconWithDataBadge(state.selectedItem?.ups, ThumbsUp)}
            {iconWithDataBadge(
              state.selectedItem?.comment_count,
              MessageSquare
            )}
            {iconWithDataBadge(state.selectedItem?.views, Eye)}
          </div>
        </div>
      )}

      <ItemModalComments comments={state.selectedItemComments} />
    </Modal>
  );
}

ItemModal.defaultProps = {
  selectedItem: {},
};

export default ItemModal;
