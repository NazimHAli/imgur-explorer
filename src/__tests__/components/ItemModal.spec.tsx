import { mockItemComments } from "@/__tests__/fixtures/mockItemComments";
import { mockItems } from "@/__tests__/fixtures/mockItems";
import { mockSelectedItem } from "@/__tests__/fixtures/mockSelectedItem";
import { fireEvent, render, screen } from "@/__tests__/fixtures/test-utils";
import ItemModal from "@/components/ItemModal";
import {
  dispatchItems,
  dispatchRequestArgs,
  dispatchSelectedItem,
} from "@/state/dispatchHelpers";
import "@testing-library/jest-dom";
import { useEffect } from "react";

const setIsOpen = jest.fn();

/* eslint-disable react/prop-types */
function TestComponent({
  setSelectedItem = false,
  setItemComments = false,
  isOpen = false,
} = {}) {
  // const state = useStore();
  useEffect(() => {
    dispatchItems(mockItems.data);

    if (setSelectedItem || setItemComments) {
      let newState = { selectedItem: mockSelectedItem };

      if (setSelectedItem) {
        dispatchRequestArgs({ selectedItemID: newState.selectedItem.id });
        // @ts-ignore
        newState = { ...newState, selectedItemComments: mockItemComments };
        dispatchSelectedItem(newState["selectedItemComments"]);
      }
    }
  }, []);

  return <ItemModal isOpen={isOpen} setIsOpen={setIsOpen} />;
}
/* eslint-enable react/prop-types */

function renderModal(args) {
  const updArgs = {
    setSelectedItem: false,
    setItemComments: false,
    isOpen: false,
    ...args,
  };

  const { setItemComments, setSelectedItem, isOpen } = updArgs;

  render(
    <TestComponent
      isOpen={isOpen}
      setItemComments={setItemComments}
      setSelectedItem={setSelectedItem}
    />,
    {
      container: document.getElementById("root"),
    }
  );
}

describe("ItemModal", () => {
  let args, testElement;

  describe("renders one element when not open", () => {
    beforeEach(() => {
      renderModal({});
    });

    test("should match snapshot", () => {
      expect(document.querySelector("body")).toMatchSnapshot();
    });
  });

  describe("when open", () => {
    beforeEach(() => {
      renderModal({ isOpen: true });
    });

    test("mounts", () => {
      expect(document.querySelector(".ReactModalPortal")).toBeDefined();
    });

    test("document.body contains modal open class", () => {
      expect(document.querySelector("body").className).toBe(
        "ReactModal__Body--open"
      );
    });

    test("images not rendered", () => {
      expect(screen.queryAllByRole("img")).toHaveLength(0);
    });

    test("has comments section", () => {
      testElement = screen.getByRole("heading", { level: 3 });
      expect(testElement).toHaveTextContent("Comments");
    });
  });

  describe("selected item", () => {
    beforeEach(() => {
      args = {
        setSelectedItem: true,
        isOpen: true,
      };
      renderModal(args);
    });

    test("img rendered", () => {
      const image = screen.queryByRole("img");
      expect(image).toMatchInlineSnapshot(`
        <img
          alt="Good night you cool cats and kittens"
          height="1536"
          loading="lazy"
          srcset="https://i.imgur.com/fLfIhwRl.jpg"
          width="2048"
        />
      `);
    });

    test("has info badges", () => {
      testElement = document.querySelectorAll("span.data-badge");
      expect(testElement).toHaveLength(3);
      mockSelectedItem;
      expect(testElement[0].dataset.count).toEqual(
        mockSelectedItem.ups.toLocaleString()
      );
      expect(testElement[1].dataset.count).toEqual(
        mockSelectedItem.comment_count.toLocaleString()
      );
      expect(testElement[2].dataset.count).toEqual(
        mockSelectedItem.views.toLocaleString()
      );
    });

    test("close modal on button click", () => {
      fireEvent.click(screen.getByRole("button"));

      testElement = document.querySelector(".item-modal");
      expect(testElement).toBeNull();
    });
  });

  describe("with item + comments", () => {
    beforeEach(() => {
      args = {
        setSelectedItem: true,
        setItemComments: true,
      };
      renderModal(args);
    });

    test("on mount sets isOpen=true", () => {
      expect(setIsOpen).toHaveBeenCalledWith(true);
    });
  });
});
