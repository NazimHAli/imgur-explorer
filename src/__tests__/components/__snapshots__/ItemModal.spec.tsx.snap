// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemModal renders one element when not open should match snapshot 1`] = `
<body>
  <div
    id="root"
  />
</body>
`;
